js/web/lib/wasm/jsep/init.ts

```diff
@@ -0,0 +1,161 @@
+
+/* eslint-disable no-bitwise */
+
+class TensorViewImpl implements TensorView {
+  constructor(
+      private module: OrtWasmModule, public readonly dataType: number, public readonly data: number,
+      public readonly dims: readonly number[]) {}
+
+  getFloat32Array(): Float32Array {
+    return new Float32Array(this.module.HEAP8.buffer, this.data, ShapeUtil.size(this.dims));
+  }
+
+  reshape(newDims: readonly number[]): TensorView {
+    if (ShapeUtil.size(newDims) !== ShapeUtil.size(this.dims)) {
+      throw new Error('Invalid new shape');
+    }
+    return new TensorViewImpl(this.module, this.dataType, this.data, newDims);
+  }
+}
+
+class OpKernelContext implements ComputeContext {
+  readonly opKernelContext: number;
+  readonly inputs: readonly TensorView[];
+  readonly customData: {[key: string]: unknown} = {};
+  constructor(private module: OrtWasmModule, private backend: WebGpuBackend, contextDataOffset: number) {
+    const heapU32 = module.HEAPU32;
+
+    // extract context data
+    let dataIndex = (contextDataOffset >> 2);
+    this.opKernelContext = heapU32[dataIndex++];
+    const inputCount = heapU32[dataIndex++];
+
+    const inputs: TensorView[] = [];
+    for (let i = 0; i < inputCount; i++) {
+      const dataType = heapU32[dataIndex++];
+      const data = heapU32[dataIndex++];
+      const dim = heapU32[dataIndex++];
+      const dims: number[] = [];
+      for (let d = 0; d < dim; d++) {
+        dims.push(heapU32[dataIndex++]);
+      }
+      inputs.push(new TensorViewImpl(module, dataType, data, dims));
+    }
+    this.inputs = inputs;
+  }
+
+  compute(program: ProgramInfoLoader|ProgramInfo, inputsOutputsMapping?: ComputeContextInputsOutputsMapping):
+      TensorView[] {
+    // prepare inputs. inputs should always be valid data.
+    const mappedInputs =
+        inputsOutputsMapping?.inputs?.map(i => typeof i === 'number' ? this.inputs[i] : i) ?? this.inputs;
+    // prepare outputs.
+    const outputIndices = inputsOutputsMapping?.outputs ?? [];
+    const createKernelOutput = (index: number, dataType: number, dims: readonly number[]): TensorView =>
+        new TensorViewImpl(this.module, dataType, this.output(index, dims), dims);
+    const createTemporaryOutput = (dataType: number, dims: readonly number[]): TensorView => {
+      const elementSize = getTensorElementSize(dataType);
+      if (!elementSize) {
+        throw new Error(`Unsupported data type: ${dataType}`);
+      }
+      const bufferSize = elementSize * ShapeUtil.size(dims);
+      return new TensorViewImpl(this.module, dataType, this.backend.gpuDataManager.create(bufferSize).id, dims);
+    };
+    return this.backend.run(program, mappedInputs, outputIndices, createKernelOutput, createTemporaryOutput);
+  }
+
+  output(index: number, dims: readonly number[]): number {
+    const stack = this.module.stackSave();
+    try {
+      const data = this.module.stackAlloc((1 + dims.length) * 4 /* sizeof(size_t) */);
```

Comment:
those are processed in C/C++ code as `uint32_t*` because we are targeting wasm32. in C/C++ code we can use `sizeof(size_t)` (and I am using it) but in ts I am doing it simple.